hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/timeline/CompletionTimeQueryView.java

```diff
@@ -0,0 +1,149 @@
+ */
+public class CompletionTimeQueryView implements AutoCloseable {
+  private final HoodieTableMetaClient metaClient;
+
+  /**
+   * Mapping from instant start time -> completion time.
+   * Should be thread-safe data structure.
+   */
+  private final Map<String, String> startToCompletionInstantTimeMap;
+
+  /**
+   * The start instant time to eagerly load from, by default load last 3 days of completed instants.
+   */
+  private final String startInstant;
+
+  /**
+   * The first instant on the active timeline, used for query optimization.
+   */
+  private final String firstInstantOnActiveTimeline;
+
+  /**
+   * The constructor.
+   *
+   * @param metaClient   The table meta client.
+   * @param startInstant The earliest instant time to eagerly load from, by default load last 3 days of completed instants.
+   */
+  public CompletionTimeQueryView(HoodieTableMetaClient metaClient, String startInstant) {
+    this.metaClient = metaClient;
+    this.startToCompletionInstantTimeMap = new ConcurrentHashMap<>();
+    this.startInstant = startInstant;
+    this.firstInstantOnActiveTimeline = metaClient.getActiveTimeline().firstInstant().map(HoodieInstant::getTimestamp).orElse("");
+    load();
+  }
+
+  /**
+   * Queries the instant completion time with given start time.
+   *
+   * @param startTime The start time.
+   *
+   * @return The completion time if the instant finished or empty if it is still pending.
+   */
+  public Option<String> getCompletionTime(String startTime) {
+    String completionTime = this.startToCompletionInstantTimeMap.get(startTime);
+    if (completionTime != null) {
+      return Option.of(completionTime);
+    }
+    if (HoodieTimeline.compareTimestamps(startTime, GREATER_THAN, this.firstInstantOnActiveTimeline)) {
+      // the instant is still pending
+      return Option.empty();
+    }
+    // the 'startTime' should be out of the eager loading range, switch to a lazy loading.
+    // This operation is resource costly.
+    HoodieArchivedTimeline.loadInstants(metaClient,
+        new EQTsFilter(startTime),
+        HoodieArchivedTimeline.LoadMode.SLIM,
+        r -> true,
+        this::readCompletionTime);
+    return Option.ofNullable(this.startToCompletionInstantTimeMap.get(startTime));
+  }
+
+  /**
+   * This is method to read instant completion time.
+   * This would also update 'startToCompletionInstantTimeMap' map with start time/completion time pairs.
+   * Only instants starts from 'startInstant' (inclusive) are considered.
+   */
+  private void load() {
+    // load active instants first.
+    this.metaClient.getActiveTimeline()
+        .filterCompletedInstants().getInstantsAsStream()
+        .forEach(instant -> setCompletionTime(instant.getTimestamp(), instant.getStateTransitionTime()));
+    // then load the archived instants.
+    HoodieArchivedTimeline.loadInstants(metaClient,
+        new HoodieArchivedTimeline.StartTsFilter(this.startInstant),
+        HoodieArchivedTimeline.LoadMode.SLIM,
```

Comment:
`SLIM` mode is sufficient? We don't need any metadata?

Reply:
Yeah, here we only want to fetch the start time and completion time for each transaction.